code/go/internal/validator/semantic/validate_required_vargroups_test.go

```diff
@@ -0,0 +1,136 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package semantic
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/yaml.v3"
+)
+
+func TestValidateRequiredVarGroups(t *testing.T) {
+	cases := []struct {
+		title    string
+		manifest string
+		errors   []string
+	}{
+		{
+			title: "good",
+			manifest: `
+vars:
+  - name: user
+  - name: password
+  - name: api_key
+policy_templates:
+  - inputs:
+    - required_vars:
+        user_password:
+          - name: user
+          - name: password
+        api_key:
+          - name: api_key
+`,
+		},
+		{
+			title: "variable defined in policy",
+			manifest: `
+vars:
+  - name: user
+  - name: password
+policy_templates:
+  - vars:
+    - name: api_key
+    inputs:
+    - required_vars:
+        user_password:
+          - name: user
+          - name: password
+        api_key:
+          - name: api_key
+`,
+		},
+		{
+			title: "missing variable",
+			manifest: `
+vars:
+  - name: user
+  - name: password
+policy_templates:
+  - inputs:
+    - required_vars:
+        user_password:
+          - name: user
+          - name: password
+        api_key:
+          - name: api_key
+`,
+			errors: []string{
+				`required var "api_key" in optional group is not defined`,
+			},
+		},
+		{
+			title: "variable defined as required",
+			manifest: `
+vars:
+  - name: user
+  - name: password
+  - name: api_key
+    required: true
+policy_templates:
+  - inputs:
+    - required_vars:
+        user_password:
+          - name: user
+          - name: password
+        api_key:
+          - name: api_key
+`,
+			errors: []string{
+				`required var "api_key" in optional group is defined as always required`,
```

Comment:
Why does it fail with error if variable api_key is defined ?

Reply:
This check is to avoid conflicts between the definition of `required` at the var level, and its presence in an optional group in `required_vars`.
A field with `required: true` would be expected to be required in any case, while a field in a `required_var` groups would be only required depending on the other groups and variables.

The test fails here because `api_key` is both, defined as `required: true` and as a member of a `required_vars` optional group.

Reply:
Got it! Thanks!